src/dump.c

```diff
@@ -1082,25 +1082,36 @@ static void jl_serialize_value_(jl_serializer_state *s, jl_value_t *v, int as_li
         }
 
         char *data = (char*)jl_data_ptr(v);
-        size_t i, np = t->layout->npointers;
-        const char *start = data;
-        for (i = 0; i < np; i++) {
-            uint32_t ptr = jl_ptr_offset(t, i);
-            jl_value_t **fld = &((jl_value_t**)data)[ptr];
-            if ((const char*)fld != start)
-                ios_write(s->s, start, (const char*)fld - start);
-            jl_value_t *e = *fld;
-            JL_GC_PROMISE_ROOTED(e);
-            if (t->mutabl && e && jl_is_cpointer(e) && jl_unbox_voidpointer(e) != (void*)-1 && jl_unbox_voidpointer(e) != NULL)
-                // reset Ptr fields to C_NULL (but keep MAP_FAILED / INVALID_HANDLE)
-                jl_serialize_cnull(s, jl_typeof(e));
-            else
-                jl_serialize_value(s, e);
-            start = (const char*)&fld[1];
-        }
-        data += jl_datatype_size(t);
-        if (data != start)
-            ios_write(s->s, start, data - start);
+        size_t i, j, np = t->layout->npointers;
+        uint32_t nf = t->layout->nfields;
+        char *last = data;
+        for (i = 0, j = 0; i < nf; i++) {
+            size_t offs = jl_field_offset(t, i);
+            char *ptr = data + offs;
+            if (t->mutabl && jl_is_cpointer_type(jl_field_type(t, i)) && *(void**)ptr != (void*)-1) {
+                if (ptr > last)
+                    ios_write(s->s, last, ptr - last);
+                char *n = NULL;
+                ios_write(s->s, (char*)&n, sizeof(n));
+                last = ptr + sizeof(n);
+            }
+            else if (j < np && ptr == (char*)&((jl_value_t**)data)[jl_ptr_offset(t, j)]) {
```

Comment:
I think this needs `while ptr > jl_ptr_offset(t, j) { /* write out previous field */ }`